### PR TITLE
Have `DataModel::Provider` be an argument to Server Init parameters

### PR DESCRIPTION
**Pull request by andy31415**
This fixes the app side for #36230

Added an `DataModel::Provider` init argument to server init params and make use of that in all main functions in our examples and code.

### CODE REVIEW

File: examples/light-switch-app/qpg/src/AppTask.cpp [hunk 1]
```diff
@@ -220,6 +222,7 @@ void AppTask::InitServer(intptr_t arg)
     VerifyOrDie(sTestEventTriggerDelegate.Init(ByteSpan(sTestEventTriggerEnableKey)) == CHIP_NO_ERROR);
     VerifyOrDie(sTestEventTriggerDelegate.AddHandler(&sFaultTestEventTriggerHandler) == CHIP_NO_ERROR);
     (void) initParams.InitializeStaticResourcesBeforeServerInit();
+    initParams.dataModelProvider        = chip::app::CodegenDataModelProviderInstance();
```
Comment (bzbarsky-apple) [inline examples/light-switch-app/qpg/src/AppTask.cpp:225]:
```suggestion
    initParams.dataModelProvider        = CodegenDataModelProviderInstance();
```

File: examples/lit-icd-app/nrfconnect/main/AppTask.cpp [hunk 2]
```diff
@@ -204,6 +205,7 @@ CHIP_ERROR AppTask::Init()
     initParams.operationalKeystore = &sPSAOperationalKeystore;
 #endif
     (void) initParams.InitializeStaticResourcesBeforeServerInit();
+    initParams.dataModelProvider        = chip::app::CodegenDataModelProviderInstance();
```
Comment (bzbarsky-apple) [inline examples/lit-icd-app/nrfconnect/main/AppTask.cpp:208]:
```suggestion
    initParams.dataModelProvider        = CodegenDataModelProviderInstance();
```

File: examples/lock-app/nrfconnect/main/AppTask.cpp [hunk 3]
```diff
@@ -229,6 +230,7 @@ CHIP_ERROR AppTask::Init()
     initParams.operationalKeystore = &sPSAOperationalKeystore;
 #endif
     (void) initParams.InitializeStaticResourcesBeforeServerInit();
+    initParams.dataModelProvider        = chip::app::CodegenDataModelProviderInstance();
```
Comment (bzbarsky-apple) [inline examples/lock-app/nrfconnect/main/AppTask.cpp:233]:
```suggestion
    initParams.dataModelProvider        = CodegenDataModelProviderInstance();
```

File: examples/lock-app/qpg/src/AppTask.cpp [hunk 4]
```diff
@@ -218,6 +220,7 @@ void AppTask::InitServer(intptr_t arg)
     VerifyOrDie(sTestEventTriggerDelegate.Init(ByteSpan(sTestEventTriggerEnableKey)) == CHIP_NO_ERROR);
     VerifyOrDie(sTestEventTriggerDelegate.AddHandler(&sFaultTestEventTriggerHandler) == CHIP_NO_ERROR);
     (void) initParams.InitializeStaticResourcesBeforeServerInit();
+    initParams.dataModelProvider        = chip::app::CodegenDataModelProviderInstance();
```
Comment (bzbarsky-apple) [inline examples/lock-app/qpg/src/AppTask.cpp:223]:
```suggestion
    initParams.dataModelProvider        = CodegenDataModelProviderInstance();
```

File: examples/ota-requestor-app/ameba/main/chipinterface.cpp [hunk 5]
```diff
@@ -23,6 +23,7 @@
 #include <DeviceInfoProviderImpl.h>
 
 #include <app/clusters/network-commissioning/network-commissioning.h>
+#include <app/codegen-data-model-provider/Instance.h>
```
Comment (bzbarsky-apple) [inline examples/ota-requestor-app/ameba/main/chipinterface.cpp:26]: Why?  Does not seem to be used in here?

File: examples/platform/infineon/cyw30739/matter_config.cpp [hunk 6]
```diff
@@ -216,6 +217,7 @@ void CYW30739MatterConfig::InitApp(void)
     // Create initParams with SDK example defaults here
     static chip::CommonCaseDeviceServerInitParams initParams;
     (void) initParams.InitializeStaticResourcesBeforeServerInit();
+    initParams.dataModelProvider        = chip::app::CodegenDataModelProviderInstance();
```
Comment (bzbarsky-apple) [inline examples/platform/infineon/cyw30739/matter_config.cpp:220]:
```suggestion
    initParams.dataModelProvider        = app::CodegenDataModelProviderInstance();
```

File: examples/platform/openiotsdk/app/openiotsdk_platform.cpp [hunk 7]
```diff
@@ -274,6 +275,7 @@ int openiotsdk_chip_run(void)
         ChipLogError(NotSpecified, "Initialize static resources before server init failed: %s", err.AsString());
         return EXIT_FAILURE;
     }
+    initParams.dataModelProvider             = chip::app::CodegenDataModelProviderInstance();
```
Comment (bzbarsky-apple) [inline examples/platform/openiotsdk/app/openiotsdk_platform.cpp:278]:
```suggestion
    initParams.dataModelProvider             = app::CodegenDataModelProviderInstance();
```

File: examples/platform/telink/common/src/AppTaskCommon.cpp [hunk 8]
```diff
@@ -301,6 +302,7 @@ CHIP_ERROR AppTaskCommon::InitCommonParts(void)
     VerifyOrDie(sTestEventTriggerDelegate.AddHandler(&sOtaTestEventTriggerHandler) == CHIP_NO_ERROR);
 #endif
     (void) initParams.InitializeStaticResourcesBeforeServerInit();
+    initParams.dataModelProvider        = chip::app::CodegenDataModelProviderInstance();
```
Comment (bzbarsky-apple) [inline examples/platform/telink/common/src/AppTaskCommon.cpp:305]:
```suggestion
    initParams.dataModelProvider        = CodegenDataModelProviderInstance();
```

File: examples/pump-app/nrfconnect/main/AppTask.cpp [hunk 9]
```diff
@@ -205,6 +206,7 @@ CHIP_ERROR AppTask::Init()
     initParams.operationalKeystore = &sPSAOperationalKeystore;
 #endif
     (void) initParams.InitializeStaticResourcesBeforeServerInit();
+    initParams.dataModelProvider        = chip::app::CodegenDataModelProviderInstance();
```
Comment (bzbarsky-apple) [inline examples/pump-app/nrfconnect/main/AppTask.cpp:209]:
```suggestion
    initParams.dataModelProvider        = app::CodegenDataModelProviderInstance();
```

File: examples/pump-controller-app/nrfconnect/main/AppTask.cpp [hunk 10]
```diff
@@ -190,6 +191,7 @@ CHIP_ERROR AppTask::Init()
     VerifyOrDie(sTestEventTriggerDelegate.Init(ByteSpan(sTestEventTriggerEnableKey)) == CHIP_NO_ERROR);
     VerifyOrDie(sTestEventTriggerDelegate.AddHandler(&sOtaTestEventTriggerHandler) == CHIP_NO_ERROR);
     (void) initParams.InitializeStaticResourcesBeforeServerInit();
+    initParams.dataModelProvider        = chip::app::CodegenDataModelProviderInstance();
```
Comment (bzbarsky-apple) [inline examples/pump-controller-app/nrfconnect/main/AppTask.cpp:194]:
```suggestion
    initParams.dataModelProvider        = CodegenDataModelProviderInstance();
```

File: examples/shell/shell_common/cmd_server.cpp [hunk 11]
```diff
@@ -58,6 +59,7 @@ static CHIP_ERROR CmdAppServerStart(int argc, char ** argv)
     // Init ZCL Data Model and CHIP App Server
     static chip::CommonCaseDeviceServerInitParams initParams;
     (void) initParams.InitializeStaticResourcesBeforeServerInit();
+    initParams.dataModelProvider             = chip::app::CodegenDataModelProviderInstance();
```
Comment (bzbarsky-apple) [inline examples/shell/shell_common/cmd_server.cpp:62]:
```suggestion
    initParams.dataModelProvider             = app::CodegenDataModelProviderInstance();
```

File: examples/window-app/nrfconnect/main/AppTask.cpp [hunk 12]
```diff
@@ -208,6 +209,7 @@ CHIP_ERROR AppTask::Init()
     initParams.operationalKeystore = &sPSAOperationalKeystore;
 #endif
     (void) initParams.InitializeStaticResourcesBeforeServerInit();
+    initParams.dataModelProvider        = chip::app::CodegenDataModelProviderInstance();
```
Comment (bzbarsky-apple) [inline examples/window-app/nrfconnect/main/AppTask.cpp:212]:
```suggestion
    initParams.dataModelProvider        = CodegenDataModelProviderInstance();
```

File: src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp [hunk 13]
```diff
@@ -14,6 +14,8 @@
  *    See the License for the specific language governing permissions and
  *    limitations under the License.
  */
+#include "app/data-model-provider/Provider.h"
+#include "app/util/persistence/AttributePersistenceProvider.h"
```
Comment (bzbarsky-apple) [inline src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp:18]: Why is the AttributePersistenceProvider include needed?

File: src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp [hunk 14]
```diff
@@ -14,6 +14,8 @@
  *    See the License for the specific language governing permissions and
  *    limitations under the License.
  */
+#include "app/data-model-provider/Provider.h"
```
Comment (bzbarsky-apple) [inline src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp:17]: Why is this in quotes, not angle brackets like our other includes?  And why is it here at all?  I'd think CodegenDataModelProvider.h includes that header, and has to....

File: src/app/codegen-data-model-provider/tests/BUILD.gn [hunk 15]
```diff
@@ -47,6 +47,8 @@ source_set("mock_model") {
     "${chip_root}/src/app/util/mock:mock_ember",
     "${chip_root}/src/lib/core:string-builder-adapters",
   ]
+
+  deps = [ "${chip_root}/src/app/util/persistence:persistence" ]
```
Comment (bzbarsky-apple) [inline src/app/codegen-data-model-provider/tests/BUILD.gn:51]: Doesn't seem like part of the actual changes in this PR.

File: src/app/server/Server.cpp [hunk 16]
```diff
@@ -161,6 +162,12 @@ CHIP_ERROR Server::Init(const ServerInitParams & initParams)
     SetAttributePersistenceProvider(&mAttributePersister);
     SetSafeAttributePersistenceProvider(&mAttributePersister);
 
+    // Ember requires a persistence provider:
```
Comment (bzbarsky-apple) [inline src/app/server/Server.cpp:165]: I don't quite understand this comment.  It's talking about "requires a persistence provider" as if that's the thing it's documenting, but it's documenting the SetDataModelProvider() call....

It seems like the actual ordering constraints here are:

1. InitDataModelHandler requires a data model provider (why?  Where does it use it?  This part is unclear to me)
2. CodegenDataModelProvider requires a persistence provider (but it does not use it directly.... the actual use likely relevant here is inside code called directly from InitDataModelHandler, not via the provider.  The provider _can_ call things that need persistence, but I don't think it does that as part of the startup sequence.)

File: src/app/server/Server.h [hunk 17]
```diff
@@ -79,7 +79,7 @@
 #include <app/icd/server/ICDCheckInBackOffStrategy.h>        // nogncheck
 #endif                                                       // CHIP_CONFIG_ENABLE_ICD_CIP
 #endif                                                       // CHIP_CONFIG_ENABLE_ICD_SERVER
-
+                                                             //
```
Comment (bzbarsky-apple) [inline src/app/server/Server.h:82]: Why this change?

File: src/app/server/Server.h [hunk 18]
```diff
@@ -192,6 +192,8 @@ struct ServerInitParams
     // If the ICD Check-In protocol use-case is supported and no strategy is provided, server will use the default strategy.
     app::ICDCheckInBackOffStrategy * icdCheckInBackOffStrategy = nullptr;
 #endif // CHIP_CONFIG_ENABLE_ICD_CIP
+
+    chip::app::DataModel::Provider * dataModelProvider = nullptr;
```
Comment (bzbarsky-apple) [inline src/app/server/Server.h:196]: This needs to be documented as "must be non-null for a valid ServerInitParams" or something like that.

File: src/app/server/Server.cpp [hunk 19]
```diff
@@ -126,6 +126,7 @@ CHIP_ERROR Server::Init(const ServerInitParams & initParams)
     VerifyOrExit(initParams.operationalKeystore != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
     VerifyOrExit(initParams.opCertStore != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
     VerifyOrExit(initParams.reportScheduler != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(initParams.dataModelProvider != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
```
Comment (bzbarsky-apple) [inline src/app/server/Server.cpp:129]: Since this will likely break every server application out there, we should have clearer logging here about what exactly failed so people can fix it.  Right now in a typical small-device config where file/line are not attached to CHIP_ERROR there will be no indication what exactly failed...

File: src/controller/python/chip/server/ServerInit.cpp [hunk 20]
```diff
@@ -161,6 +162,7 @@ PyChipError pychip_server_native_init()
     // Init ZCL Data Model and CHIP App Server
     static chip::CommonCaseDeviceServerInitParams initParams;
     PyReturnErrorOnFailure(ToPyChipError(initParams.InitializeStaticResourcesBeforeServerInit()));
+    initParams.dataModelProvider             = chip::app::CodegenDataModelProviderInstance();
```
Comment (bzbarsky-apple) [inline src/controller/python/chip/server/ServerInit.cpp:165]:
```suggestion
    initParams.dataModelProvider             = app::CodegenDataModelProviderInstance();
```

File: src/app/server/Server.cpp (hunk 19 above)
Comment (andy31415) [reply src/app/server/Server.cpp:129]: Added logs as well as a note in our upgrading document.

File: src/app/server/Server.cpp (hunk 16 above)
Comment (andy31415) [reply src/app/server/Server.cpp]: Updated the comment. We essentially want to move all ember specific init in the data model provider since that is our "interface to ember". However until then, the ordering in Server.cpp is very intentional.

The call stack is generally "load defaults" which may load persisted NVM values it seems:

```
InitDataModelHandler -> emberAfInit -> emberAfInitializeAttributes -> emAfLoadAttributeDefaults
```

File: src/app/server/Server.cpp [hunk 21]
```diff
@@ -162,10 +170,18 @@ CHIP_ERROR Server::Init(const ServerInitParams & initParams)
...
+    // early enough.
+    // In particular ember initialization (emberAfInit) is done in InitDataModelHandler which happens
+    // below and may load NVM attributes, so data model should be initialized (e.g. as we transition to
+    // having data model provider initialization starting up ember prerequisites). The call stack is:
+    //    - InitDataModelHandler -> emberAfInit -> emberAfInitializeAttributes -> emAfLoadAttributeDefaults
+    //      may load from persistent storage.
+    //
+    // As we expect the CodegenDatamodelProvider to everntually be responsible for ember attribute
+    // persistence initialization, we place this:
+    //   - After SetSefeAttributePersistenceProvider (as this is the global persistence provider)
```
Comment (bzbarsky-apple) [inline src/app/server/Server.cpp:183]:
```suggestion
    //   - After SetSafeAttributePersistenceProvider (as this is the global persistence provider)
```

File: src/app/server/Server.cpp [hunk 22]
```diff
@@ -162,10 +170,18 @@ CHIP_ERROR Server::Init(const ServerInitParams & initParams)
...
+    //   - Before InitDataModelHandler to allow data model startup to set up persistence for ember as needed
```
Comment (bzbarsky-apple) [inline src/app/server/Server.cpp:184]: OK, but the plan is presumably to have InitDataModelHandler go away altogether, because literally the only thing it does is start up the ember bits, right?

File: src/app/server/Server.cpp [hunk 23]
```diff
@@ -162,10 +170,18 @@ CHIP_ERROR Server::Init(const ServerInitParams & initParams)
...
-    //     enough, after SetSafeAttributePersistenceProvider
+    // Until all ember calls are handled by the data model provider ensure that this is done
+    // early enough.
+    // In particular ember initialization (emberAfInit) is done in InitDataModelHandler which happens
+    // below and may load NVM attributes, so data model should be initialized (e.g. as we transition to
+    // having data model provider initialization starting up ember prerequisites). The call stack is:
+    //    - InitDataModelHandler -> emberAfInit -> emberAfInitializeAttributes -> emAfLoadAttributeDefaults
+    //      may load from persistent storage.
+    //
+    // As we expect the CodegenDatamodelProvider to everntually be responsible for ember attribute
```
Comment (bzbarsky-apple) [inline src/app/server/Server.cpp:181]:
```suggestion
    // As we expect the CodegenDatamodelProvider to eventually be responsible for ember attribute
```

File: src/app/server/Server.cpp (hunk 21 above)
Comment (bzbarsky-apple) [inline src/app/server/Server.cpp:183]: But **SetDataModelProvider** is not going to be responsible for "ember attribute persistence initialization"....  This comment and logic are still not making sense to me.  In that future world in which the codegen data model provider is responsible for persistence, it will be calling SetAttributePersistenceProvider itself, in its init, way before this code, no?

File: src/app/server/Server.cpp [hunk 24]
```diff
@@ -161,6 +170,20 @@ CHIP_ERROR Server::Init(const ServerInitParams & initParams)
...
+    // persistence initialization, we place this:
+    //   - After SetSefeAttributePersistenceProvider (as this is the global persistence provider)
+    //   - Before InitDataModelHandler to allow data model startup to set up persistence for ember as needed
```
Comment (bzbarsky-apple) [inline src/app/server/Server.cpp:184]:
```suggestion
    // SetDataModelProvider() actually initializes/starts the provider.  We need
    // to preserve the following ordering guarantees:
    //
    // 1) Provider initialization (under SetDataModelProvider) happens after
    //    SetSafeAttributePersistenceProvider, since the provider can then use
    //    the safe persistence provider to implement and initialize its own attribute persistence logic.
    // 2) For now, provider initialization happens before InitDataModelHandler(), which depends
    //    on atttribute persistence being already set up before it runs.  Longer-term, the logic from
    //    InitDataModelHandler should just move into the codegen provider.
```

File: src/app/server/Server.cpp (hunk 22 above)
Comment (andy31415) [reply src/app/server/Server.cpp]: Yes, all ember functions should only ever be used or seen by the codegen model. Ember is a possible implementation of the data model provider.

**Review by bzbarsky-apple — APPROVED** [review]

File: examples/light-switch-app/qpg/src/AppTask.cpp [hunk 25]
```diff
@@ -204,6 +205,7 @@ void AppTask::InitServer(intptr_t arg)
 {
     static chip::CommonCaseDeviceServerInitParams initParams;
     (void) initParams.InitializeStaticResourcesBeforeServerInit();
+    initParams.dataModelProvider = CodegenDataModelProviderInstance();
```
[the remark this answers is not in the document]
Comment (andy31415) [reply examples/light-switch-app/qpg/src/AppTask.cpp:208]: Good catch - this was via an automated script and apparently https://github.com/project-chip/connectedhomeip/blob/master/examples/light-switch-app/qpg/src/AppTask.cpp#L222 does 2 times initialization for `InitializeStaticResourcesBeforeServerInit`. Will remove one.

[the remark this answers is not in the document]
Comment (andy31415) [reply examples/light-switch-app/qpg/src/AppTask.cpp]: Updated. I am also unclear why the initParams is static here, seems wrong...
This probably needs some cleanups overall, however that should be probably by some qpg maintainer.

**Review by tehampson — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]
